### PR TITLE
Fix atime handling and relatime

### PR DESCRIPTION
**Pull request by tuxoko**
The problem for atime:

We have 3 places for atime: inode->i_atime, znode->z_atime and SA. And its
handling is a mess. A huge part of mess regarding atime comes from
zfs_tstamp_update_setup, zfs_inode_update, and zfs_getattr, which behave
inconsistently with those three values.

zfs_tstamp_update_setup clears z_atime_dirty unconditionally as long as you
don't pass ATTR_ATIME. Which means every write(2) operation which only updates
ctime and mtime will cause atime changes to not be written to disk.

Also zfs_inode_update from write(2) will replace inode->i_atime with what's
inside SA(stale). But doesn't touch z_atime. So after read(2) and write(2).
You'll have i_atime(stale), z_atime(new), SA(stale) and z_atime_dirty=0.

Now, if you do stat(2), zfs_getattr will actually replace i_atime with what's
inside, z_atime. So you will have now you'll have i_atime(new), z_atime(new),
SA(stale) and z_atime_dirty=0. These will all gone after umount. And you'll
leave with a stale atime.

The problem for relatime:

We do have a relatime config inside ZFS dataset, but how it should interact
with the mount flag MS_RELATIME is not well defined. It seems it wanted
relatime mount option to override the dataset config by showing it as
temporary in `zfs get`. But at the same time, `zfs set relatime=on|off` would
also seems to want to override the mount option. Not to mention that
MS_RELATIME flag is actually never passed into ZFS, so it never really worked.

How Linux handles atime:

The Linux kernel actually handles atime completely in VFS, except for writing
it to disk. So if we remove the atime handling in ZFS, things would just work,
no matter it's strictatime, relatime, noatime. And whenever VFS updates the
i_atime, it will notify the underlying filesystem via sb->dirty_inode().

And also there's one thing to note about atime flags like MS_RELATIME and
other flags like MS_NODEV, etc. They are mount point flags rather than
filesystem(sb) flags. Since native linux filesystem can be mounted at multiple
places at the same time, they can all have different atime settings. So these
flags are never passed down to filesystem drivers.

What this patch tries to do:

We remove znode->z_atime, since we won't gain anything from it. We remove most
of the atime handling and leave it to VFS. The only thing we do with atime is
to write it when dirty_inode() or setattr() is called. We also add
file_accessed() in zpl_read() since it's not provided in vfs_read().

### CODE REVIEW

**Comment by rlaager** [comment]
What's the status on merging inodes and znodes, or at least eliminating the duplicated fields?

**Comment by tuxoko** [comment]
@rlaager 
I don't think I've seen anyone working on it yet.

**Comment by tuxoko** [comment]
Strange, atime_002_neg keeps failing. But I haven't failed it on my box.

**Comment by spacelama** [comment]
On Fri, 1 Apr 2016, tuxoko wrote:

> What this patch tries to do:
>
> We remove znode->z_atime, since we won't gain anything from it. We remove most
> of the atime handling and leave it to VFS. The only thing we do with atime is
> to write it when dirty_inode() or setattr() is called. We also add
> file_accessed() in zpl_read() since it's not provided in vfs_read().
> You can view, comment on, or merge this pull request online at:

Thankyou!  I had been thinking atimes were broken for a while despite my
earlier bug about atime/relatime handling.  Never had time to look into
(still don't have time to test :( )


-- 
Tim Connors

File: module/zfs/zfs_sa.c [hunk 1]
```diff
@@ -309,6 +309,7 @@ zfs_sa_upgrade(sa_handle_t *hdl, dmu_tx_t *tx)
 
 	/* First do a bulk query of the attributes that aren't cached */
 	bulk = kmem_alloc(sizeof (sa_bulk_attr_t) * 20, KM_SLEEP);
```
Comment (behlendorf) [inline module/zfs/zfs_sa.c]: Not a defect in this patch (or necessarily in the existing code), but it's interesting that bulk is already oversized here.

File: module/zfs/zfs_znode.c [hunk 2]
```diff
@@ -546,8 +546,6 @@ zfs_znode_alloc(zfs_sb_t *zsb, dmu_buf_t *db, int blksz,
 	    &zp->z_pflags, 8);
 	SA_ADD_BULK_ATTR(bulk, count, SA_ZPL_PARENT(zsb), NULL,
 	    &parent, 8);
-	SA_ADD_BULK_ATTR(bulk, count, SA_ZPL_ATIME(zsb), NULL,
-	    &zp->z_atime, 16);
```
Comment (behlendorf) [inline module/zfs/zfs_znode.c]: `sa_bulk_attr_t bulk[9];` should be changed to 8.

File: module/zfs/zfs_znode.c [hunk 3]
```diff
@@ -1199,8 +1197,6 @@ zfs_rezget(znode_t *zp)
 	    &zp->z_links, sizeof (zp->z_links));
 	SA_ADD_BULK_ATTR(bulk, count, SA_ZPL_FLAGS(zsb), NULL,
 	    &zp->z_pflags, sizeof (zp->z_pflags));
-	SA_ADD_BULK_ATTR(bulk, count, SA_ZPL_ATIME(zsb), NULL,
-	    &zp->z_atime, sizeof (zp->z_atime));
```
Comment (behlendorf) [inline module/zfs/zfs_znode.c]: `sa_bulk_attr_t bulk[8];` should be changed to 7.

File: module/zfs/zfs_znode.c [hunk 4]
```diff
@@ -1346,27 +1308,9 @@ zfs_tstamp_update_setup(znode_t *zp, uint_t flag, uint64_t mtime[2],
 {
 	timestruc_t	now;
 
-	ASSERT(have_tx == !(flag & ATTR_ATIME));
```
Comment (behlendorf) [inline module/zfs/zfs_znode.c]: `have_tx` is now entirely unused, I'd suggest removing the argument entirely to avoid any confusion (from the comment too).

File: module/zfs/zfs_znode.c [hunk 5]
```diff
@@ -1296,49 +1292,15 @@ zfs_compare_timespec(struct timespec *t1, struct timespec *t2)
...
-
-	return (B_FALSE);
-}
-
-/*
  * Prepare to update znode time stamps.
  *
  *	IN:	zp	- znode requiring timestamp update
- *		flag	- ATTR_MTIME, ATTR_CTIME, ATTR_ATIME flags
+ *		flag	- ATTR_CTIME, ATTR_ATIME flags
```
Comment (behlendorf) [inline module/zfs/zfs_znode.c]: You dropped the wrong flag here accidentally, only ATTR_CTIME and ATTR_MTIME are understood.  This comment block would be a good place to add a short comment about why atime is handled differently.

File: module/zfs/zpl_file.c [hunk 6]
```diff
@@ -277,6 +278,7 @@ zpl_iter_read_common(struct kiocb *kiocb, const struct iovec *iovp,
 	    nr_segs, &kiocb->ki_pos, seg, filp->f_flags, cr, skip);
 	crfree(cr);
 
+	file_accessed(filp);
```
Comment (behlendorf) [inline module/zfs/zpl_file.c]: We _may_ need to do the same thing for mmap if the VFS doesn't handle this for us with a dirty inode call.  All that's strictly required is atime be update sometime between mmap() and munmap().

File: module/zfs/zfs_znode.c [hunk 7]
```diff
@@ -653,7 +655,12 @@ zfs_inode_update(znode_t *zp)
 	ip->i_blkbits = SPA_MINBLOCKSHIFT;
 	ip->i_blocks = i_blocks;
 
-	ZFS_TIME_DECODE(&ip->i_atime, atime);
+	/*
+	 * Only read atime if we are new inode (or rezget), otherwise i_atime
+	 * might be dirty.
```
Comment (behlendorf) [inline module/zfs/zfs_znode.c]: I'd suggest rewording this so it's clear that "new inode" really means "newly created inode read from disk".

File: module/zfs/zfs_znode.c [hunk 8]
```diff
@@ -620,8 +622,8 @@ zfs_set_inode_flags(znode_t *zp, struct inode *ip)
  * inode has the correct field it should be used, and the ZFS code
  * updated to access the inode.  This can be done incrementally.
  */
-void
-zfs_inode_update(znode_t *zp)
+static void
+__zfs_inode_update(znode_t *zp, int new)
```
Comment (behlendorf) [inline module/zfs/zfs_znode.c]: The following would be more consistent with the rest of the code base.

```
zfs_inode_update_impl(znode_t *zp, boolean_t new)
```

File: module/zfs/zfs_znode.c [hunk 9]
```diff
@@ -492,6 +492,8 @@ zfs_inode_set_ops(zfs_sb_t *zsb, struct inode *ip)
 	}
 }
 
+static void zfs_inode_update_new(znode_t *zp);
```
Comment (behlendorf) [inline module/zfs/zfs_znode.c]: As a general rule forward declarations are discouraged if it's easy to reorder the functions to avoid it.  Please go ahead and just move the `zfs_inode_update_impl` function up and wrappers up.

**Comment by behlendorf** [comment]
@rlaager I'm not aware of anyone working on it right now.  But it definitely needs to happen.

@tuxoko the approach taken in the patch looks good to me, I made a few fairly minor comment inline.  In addition, I strongly suggest adding a few new test cases to `tests/zfs-tests/tests/functional/atime/` which verify that the Linux atime mount options are set properly and behave as expected.

I'm glad to see this long standing bit of terribleness finally get addressed!

File: module/zfs/zpl_file.c (hunk 6 above)
Comment (tuxoko) [inline module/zfs/zpl_file.c]: generic_file_mmap() has file_accessed.

File: module/zfs/zfs_znode.c (hunk 2 above)
Comment (tuxoko) [inline module/zfs/zfs_znode.c]: Done.

File: module/zfs/zfs_znode.c (hunk 3 above)
Comment (tuxoko) [inline module/zfs/zfs_znode.c]: Done.

File: module/zfs/zfs_znode.c (hunk 4 above)
Comment (tuxoko) [inline module/zfs/zfs_znode.c]: Done.

File: module/zfs/zfs_znode.c (hunk 5 above)
Comment (tuxoko) [inline module/zfs/zfs_znode.c]: Done.

File: module/zfs/zfs_znode.c (hunk 7 above)
Comment (tuxoko) [inline module/zfs/zfs_znode.c]: Done.

File: module/zfs/zfs_znode.c (hunk 8 above)
Comment (tuxoko) [inline module/zfs/zfs_znode.c]: Done.

File: module/zfs/zfs_znode.c (hunk 9 above)
Comment (tuxoko) [inline module/zfs/zfs_znode.c]: Done.

**Comment by tuxoko** [comment]
Update:
- Fix thing according to review.
- Remove ZFS_ACCESSTIME_STAMP in zfs_getpage
- Fix MNTOPT_LAZYTIME comment.

**Comment by behlendorf** [comment]
Great, it looks like the only thing remaining is to add a test case.  It looks like it would be trivial to add a  `atime_003_pos.ksh` test case which verifies the `relatime` property using `atime_001_pos.ksh` as a base.  The helper functions already exist to easy check the atime and verify it.

**Comment by tuxoko** [comment]
Update: add atime_003_pos to test relatime.

**Comment by behlendorf** [comment]
This LGTM, thanks for making those fixes and adding the test case.  I see you also were able to drop ZFS_ACCESSTIME_STAMP entirely which is great.  @tuxoko if your happy with this as a final version let me know and I'll merge it.

**Comment by tuxoko** [comment]
Yeah, I'm good as it is now.

**Comment by behlendorf** [comment]
Merged as:

2b54cb1 Add zfs-tests for relatime
6760077 Make zfs mount according to relatime config in dataset
704cd07 Enable lazytime semantic for atime
0df9673 Fix atime handling and relatime

**Comment by ThomasWaldmann** [comment]
So, does this PR fix this?: https://github.com/borgbackup/borg/issues/1172

Will this fix somehow get into what ubuntu has in 16.04LTS?

**Comment by tuxoko** [comment]
@ThomasWaldmann 
Yes, it should work with O_NOATIME. Unfortunately, this patch set is not included in 0.6.5 stable branch.

**Comment by ThomasWaldmann** [comment]
OK, at least it is fixed. Somewhere.

**Comment by behlendorf** [comment]
This is something we could consider back porting to 0.6.5.

**Comment by tuxoko** [comment]
@behlendorf 
Can you queue this in the 0.6.5.9 project, thanks.

**Comment by behlendorf** [comment]
I've added it to the [0.6.5.9 project](https://github.com/zfsonlinux/zfs/projects/4) for @nedbass's consideration.
